Reduce indexing during initial update_index

calum-chamberlain opened the pull request:
`EventBank.update_index` iterates through the directory structure twice, once to count the number of files and the second to generate the index. For large databases the iteration can be a large time-sink.  In this PR an optional argument of `num_files` is added as a work-around to stop the initial iteration and get straight into the bulk of the work. This is a slightly clumsy way to do this and I'm very open to other better thought out options.

It also fixes a minor typo in the doc-string, and passes `skip_hidden` through the iterator.

Annoyingly I ran black after I added and the black patch din't come through - extra commit to come.

Just as a note: obsplus is cool. I really like the type hinting and seeing nice code. Good job. I'm excited to use this for database management for real-time matched-filter detection. Thank you @d-chambers

d-chambers commented:
Hey @calum-chamberlain, thanks for the fixes. Azure is failing on something unrelated. Dont worry about the number of commits; we have taken to squashing and merging PRs for a clean git history. 

Can you help me understand your use case a bit better? Are you creating events in batches, saving to disk, then updating the index? Maybe there could be a better way to update the index based on the events in memory and the anticipated path for each, then you could save the expense of scanning any files or reading any xmls. In my experience iterating the files is relatively fast, and there might a better fix than just skipping one of directory scans. 

Also, I would be keen to get your feedback on the `DataSet` object and its potential for sharing seismological datasets. In particular take a look at the [template](www.github.com/seismopy/opsdata) for how I would expect others to use it. We had talked awhile back about the usefulness of "reference" datasets for bench-marking detection methods and I think this would make it possible.

calum-chamberlain commented:
The use case is ridiculous and if this hack isn't desirable then you should ignore it! The use case is as follows:
1. Acquired the entire New Zealand national earthquake catalogue between 1987 and 2018 (~587,000 events);
2. All events are stored in `{year}/{month}/{event_id_short}` directories as a quakeml alongside cut miniseed files (with different processing applied to each miniseed file), a json version of the quakeml and some other files (links to other events for repeating earthquake work);
3. Realise I need a database manager for another project (real-time aftershock detection using matched-filters) which needs to be able to select earthquakes within regions and get all of the relevant information;
4. Start writing something then realise everything I need is already in obsplus (apart from a few methods to create template waveforms - I plan on subclassing `EventBank` to add a few methods in [RT_EQcorrscan](https://github.com/eqcorrscan/RT_EQcorrscan/blob/master/rt_eqcorrscan/core/database_manager.py));
5. Index the whole database of half a million events from scratch with `EventBank`

Iterating through the files is definitely fast, but still adds a few hours to the indexing process. The indexing took about 26 hours without the initial iteration. 

The `DataSet` template looks interesting, I will have a look in more detail next week!

d-chambers commented:
>  Iterating through the files is definitely fast, but still adds a few hours to the indexing process

There may be something else going on here. We have a continuous data archive with a few  million files. Using `WaveBank` (which uses the same `iter_files` function to find what needs indexing) and it takes < 5 seconds sum all the files.

in ipython can you try this:

>>> %time a = sum([1 for _ in obsplus.utils.iter_files('path/to/ridiculous/directory')] )

Also what operating system are you using?

calum-chamberlain commented:
:+1: Timing at the moment, it is definitely taking a while: this is using a network drive on a cluster computer though - so it might be limited by the ethernet connection. I also don't know if the RAID configuration would slow things down (for one I don't know what RAID it is running, or anything about the drive speed).  The connection is a GB ethernet connection over LAN.

OS wise, the cluster is running Ubuntu 16.04. Not much I can do about that. Any thoughts on how OS would affect this?

Good to know that my dataset isn't too ridiculous!

calum-chamberlain commented:
Connection to cluster timed out before it finished - it had been running for about 2 hours without a result.

d-chambers commented:
Wow! Sorry for doubting, that is really strange. I wonder if it has to move the entire file over the network  or maybe it is just latency stacking up on every single file count? I asked about the OS because I think scandir is making OS specific calls, so it could be faster on some OS compared to others, but not by that much.

So the way you are using the `num_files` argument is just to disable the first iteration and thus the progress bar true? Since this is probably not a common use case I would propose making a class attribute in the base 'Bank' class called 'use_progress_bar' or something along those lines. Then you would just use it like so:

```python
import obsplus

ebank = obsplus.EventBank('path/to/dir')
ebank.use_progress_bar = False
ebank.update_index()
```

That way the `update_index` method wont get cluttered up with seldom used arguments, but it will still be possible to do what you need. 

You are welcome to take a stab at it or I can do it in the next few days if you don't have time. 

I once created a package for storing event data in MongoDB, but I never open-sourced it because I don't really want to maintain it. If you think it might help you I am happy to share it.

calum-chamberlain commented:
Agree on not cluttering with seldom used arguments.  I added a class attribute to `Bank` and cleaned out the `num_files` argument. I was using the `num_files` to keep a progress bar but with a known number of files, but it wasn't necessary.

I haven't documented the class attribute well - not sure how these should be documented?

d-chambers commented:
> I haven't documented the class attribute well - not sure how these should be documented?

I guess we should just follow numpy docstring guidelines and have an " Attributes" section. 

I am also thinking we should remove the `bar` and `min_files_for_bar` parameters in favor of class attributes as well because they probably fall under the "seldom used" category. Since obsplus is still under 0.1.0 I don't feel bad for breaking changes that likely wont effect anyone. 

Do you mind if I push a few commits related to `update_index` to your branch? It might be cleaner to just have them in one PR.

calum-chamberlain commented:
Go for it!

CJ Chamberlain, out of office

________________________________
From: Derrick <notifications@github.com>
Sent: Wednesday, June 26, 2019 5:22:36 AM
To: niosh-mining/obsplus
Cc: Calum Chamberlain; Mention
Subject: Re: [niosh-mining/obsplus] Reduce indexing during initial update_index (#106)


I haven't documented the class attribute well - not sure how these should be documented?

I guess we should just follow numpy docstring guidelines and have an " Attributes" section.

I am also thinking we should remove the bar and min_files_for_bar parameters in favor of class attributes as well because they probably fall under the "seldom used" category. Since obsplus is still under 0.1.0 I don't feel bad for breaking changes that likely wont effect anyone.

Do you mind if I push a few commits related to update_index to your branch? It might be cleaner to just have them in one PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fniosh-mining%2Fobsplus%2Fpull%2F106%3Femail_source%3Dnotifications%26email_token%3DACTIM44Y2LUVKSWSFLN2CJDP4JH5ZA5CNFSM4H2LJJXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYQ7C7I%23issuecomment-505540989&data=02%7C01%7Ccalum.chamberlain%40vuw.ac.nz%7C467caa7f4fd4497231c608d6f991b916%7Ccfe63e236951427e8683bb84dcf1d20c%7C0%7C0%7C636970801649446152&sdata=3cOdJG8AEUHOZChmp451kq2EDC9bwZVlRwJfAJb7TDQ%3D&reserved=0>, or mute the thread<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTIM4YIE2JD2GTJCYYF5SDP4JH5ZANCNFSM4H2LJJXA&data=02%7C01%7Ccalum.chamberlain%40vuw.ac.nz%7C467caa7f4fd4497231c608d6f991b916%7Ccfe63e236951427e8683bb84dcf1d20c%7C0%7C0%7C636970801649446152&sdata=2WEGPTBA%2Fk6xIyYjgN5mlNgQQt%2FwUSuuYc%2BT8dX%2B9aA%3D&reserved=0>.

d-chambers commented:
Ok,

Sorry for being indecisive on this one but I think I figured out a better solution.

Rather than using a class attribute, we can just init a bar instance with the known number of files and use that without counting the files. That way you can avoid the file count and also use the progress bar. 

```python
import obsplus

ebank = obsplus.EventBank('path/to/directory') 
# create a progress bar instance
bar = obsplus.utils.get_progressbar(num_files=1_000_000)
# use it to update the index
ebank.update_index(bar=bar)
```

d-chambers commented:
Ok, everything is green, will squash and merge.
